### PR TITLE
Make exclude_folders and exclude_projects use sets and use for_each to avoid unnecessary resource updates

### PR DESCRIPTION
**Pull request by MajorBreakfast**
Currently, when adding new excluded folders or projects in to the middle of the list, Terraform shows unnecessary changes due to index shifts. With this change, this no longer happens. This makes it possible to order the exclusions alphabetically without getting unnecessary changes in the plan.

### CODE REVIEW

**Comment by MajorBreakfast** [comment]
Could you provide me the details why this does not pass? I don't have access

**Comment by MajorBreakfast** [comment]
@bharathkkb could you maybe have a look again? ^^'

**Comment by bharathkkb** [comment]
@MajorBreakfast thanks for the contribution

It is failing due to a problem on our end that we are working to resolve. @morgante PTAL

**Comment by MajorBreakfast** [comment]
Any progress on this? The reshuffling makes it so that we get 409 errors whenever we add an exclusion to the middle of the list. That's inconvenient

File: variables.tf [hunk 1]
```diff
@@ -55,14 +55,14 @@ variable "deny" {
...
+  type        = set(string)
   default     = [""]
 }
 
 variable "exclude_projects" {
-  description = "List of projects to exclude from the policy"
-  type        = list(string)
+  description = "Set of projects to exclude from the policy"
+  type        = set(string)
   default     = [""]
```
Comment (bharathkkb) [inline variables.tf]: for_each will detect this default [""] as 1 element and will proceed  to create` resource "google_project_organization_policy" "project_policy_list_exclude_projects"` with `""` as project which will fail when we are calling only `exclude_folders`.[ Example invocation where failure happens]( https://github.com/terraform-google-modules/terraform-google-org-policy/blob/master/examples/boolean_org_exclude/main.tf)

**Comment by bharathkkb** [comment]
@MajorBreakfast CI problems on our end have been resolved and since tests were not passing I tried debugging- left some comments with my findings. 
you can execute the tests locally as described [here](https://github.com/terraform-google-modules/terraform-google-org-policy/blob/master/CONTRIBUTING.md#integration-testing)

File: variables.tf (hunk 1 above)
Comment (MajorBreakfast) [reply variables.tf]: Thanks. I would have expected this to already fail before, though

**Comment by MajorBreakfast** [comment]
Tests are green now

**Comment by bharathkkb** [comment]
@MajorBreakfast LGTM overall. nit: I assume the locals like exclude_folders_list_length are not used anymore? They are still in main.tf

File: variables.tf (hunk 1 above)
Comment (bharathkkb) [reply variables.tf]: It was not previously failing due to the way we were calculating length
`length(compact(var.exclude_folders))`. Compact got rid of those empty `""` which then gave a length of zero

**Comment by MajorBreakfast** [comment]
Yep. The search confirms that they‘re unused. I removed them

**Comment by MajorBreakfast** [comment]
@morgante can you have a look?

**Comment by MajorBreakfast** [comment]
@bharathkkb ping

**Comment by bharathkkb** [comment]
@morgante PTAL

**Comment by MajorBreakfast** [comment]
@morgante The old code computed the length via `length(compact(var.exclude_folders))` which means the length was also computed before. Are you sure that this change could break in more scenarios? And if so, could you provide an example?

**Comment by MajorBreakfast** [comment]
@morgante This PR might have a problem. The size of the set depends on how many unique values there are. This means the size cannot be known statically if one or more entires are only known dynamically (because they're a resource attribute). This makes the value not allowd in a for_each. I have look deeply into this, but it I just came across a similar scenario and wanted to mention it here.

**Comment by morgante** [comment]
Yeah that's unfortunately true. I don't think there's any easy fix though.

**Pull request merged**